### PR TITLE
Fix client refresh and url redirection

### LINKED ISSUE #279: PASS URLs should not contain # symbol
**Describe the bug**
Currently, PASS URLS that require a login to access include a `#` in their path. However, this makes them invalid targets of redirect in certain libraries. The `#` in a URI is used to indicate a [Fragment Identifier](https://www.w3.org/Addressing/URL/4_URI_Recommentations.html) , a certain place within a page. It should not be used to identify a page itself.

**Expected behavior**
We can update the URL paths to remove  to make them simpler. Ideally we should be able to just have things like:

- www.pass.com
- www.pass.com/login
- www.pass.com/clients
- www.pass.com/documents
- www.pass.com/messages

### PR DESCRIPTION
**Pull request by timbot1789**
The goal of this PR is to remove the `#` parameter in the PASS url so we are able to use server-side redirects. These redirects will make it easier to have multiple different login flows (e.g. existing vs new user). Along the way we rebuilt and simplified the login and routing flow of the app.

I have a production build of this running up at https://279-remove-hash-from-url--splendorous-nougat-3f0062.netlify.app 

That should let you test prod behavior without difficulty.

A few changes in this PR:

1. Removed `@inrupt/solid-ui-react` and added the components we use as native code. We really only use this library as a convenient way to manage the session provided by `solid-client-authn-browser`. However, that auth library is fairly buggy, and we need to implement a few workarounds to get it running properly. Having it hidden behind a react library makes that difficult.
2. Updates `solid-client-authn-browser` to get a fix for `restorePreviousSession`. This allows the solid client to handle session restoration for us, and makes the silent authentication that happens on page refresh faster. This also made some code redundant, so I removed that. Less code ≈ fewer bugs.
3. Added code in `vite.config.js` to force vite to build `solid-client` as a ES module instead of a CommonJS bundle. This is a workaround for this bug: https://github.com/inrupt/solid-client-authn-js/issues/3001
4. Renamed the `routes` folder to `pages` to keep it consistent with how the files are referred to internally.
5. Removes the hash router and replaces it with browser router. This is to enable server redirection and other features used by PASS login. The main use case for this right now is in user registration. To register a user, we need them to log into a pod. That requires sending a redirect URL to the pod server. If the user registration page is hosted on `www.pass.com/PASS/#/signup`, the redirect link could be `www.pass.com/PASS/#/signup`, and when we get back to that page with a valid login, we continue with any  user registration flow. However, URLs with a `#` in them are not valid redirect links. Sending such a link to the server will cause an error. If we instead put signup at `www.pass.com/signup`, we can redirect to it without issue after we've finished logging into the pod.
 
NOTE: This will break github code pages, but we can deploy the project to another site without too much difficulty, We will need to do that at some point anyway.

6. Updated routes so that they don't have `#` in them anymore. Routes in this PR are:
  - `www.pass.com` - login/landing page
  - `www.pass.com/clients`
  - `www.pass.com/documents`
  - `www.pass.com/messages`
  - `www.pass.com/profile`
7. Added mappings of a lot of the `src` modules in `vite.config.js` so they have absolute references. This makes it easier to refer to them throughout the project, since they will always have the same path no matter what folder you're in.
 ---
No new unit tests because this doesn't implement a new feature, just refactors. The fact that there's no regression in the unit tests should demonstrate that the refactor is working correctly.

We could add some unit tests for routes, but that can be in a future PR.

### CODE REVIEW

File: src/pages/Documents.jsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
+// React Imports
```
Comment (timbot1789) [inline src/pages/Documents.jsx:1]: This document wasn't created. It was moved from /routes/Document.jsx. Only new feature is the layout wrapping it.

File: test/components/NavBar/__snapshots__/OidcLoginComponent.test.jsx.snap [hunk 2]
```diff
@@ -33,6 +33,7 @@ exports[`renders correctly 1`] = `
     class="MuiBox-root css-120gsus"
   />
   <div>
+    http://localhost:3000/PASS/
```
Comment (timbot1789) [inline test/components/NavBar/__snapshots__/OidcLoginComponent.test.jsx.snap:36]: We now pass the redirect URL in explicitly, causing it to render correctly in the test (it wasn't before, and I didn't notice).

File: test/pages/__snapshots__/Home.test.jsx.snap [hunk 3]
```diff
@@ -0,0 +1,280 @@
+// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html
```
Comment (timbot1789) [inline test/pages/__snapshots__/Home.test.jsx.snap:1]: Moved, not created

File: src/components/NavBar/OidcLoginComponent.jsx [hunk 4]
```diff
@@ -4,14 +4,12 @@ import React, { useState } from 'react';
...
 import { TextField, Box, Button } from '@mui/material';
-// Custom Hook Imports
-import { useRedirectUrl } from '../../hooks';
 // Constants Imports
 import { ENV } from '../../constants';
 
 const OidcLoginComponent = () => {
-  const [oidcIssuer, setOidcIssuer] = useState(ENV.VITE_SOLID_IDENTITY_PROVIDER);
-  const redirectUrl = useRedirectUrl();
+  const defaultOidc = ENV.VITE_SOLID_IDENTITY_PROVIDER || '';
```
Comment (timbot1789) [inline src/components/NavBar/OidcLoginComponent.jsx:12]: We were getting a controlled component error in dev console if the ENV didn't set this value. This guard cleans up the error.

File: test/components/NavBar/__snapshots__/OidcLoginComponent.test.jsx.snap [hunk 5]
```diff
@@ -82,6 +83,7 @@ exports[`sets OIDC provider on login 1`] = `
   />
   <div>
     oidc.provider.url
+    http://localhost:3000
```
Comment (timbot1789) [inline test/components/NavBar/__snapshots__/OidcLoginComponent.test.jsx.snap]: We now pass the redirect URL in explicitly, causing it to render correctly in the test (it wasn't before, and I didn't notice).

**Comment by xscottxbrownx** [comment]
I haven't looked at this at all yet, but if it does anything to help the redirects or logged in ? check - then I'm all for it. Constantly get frustrated with getting stuck at splash/intro page of PASS and nothing happens when clicking `login`.    I mean I never log out of solid, it should be able to see that I am logged into solid and bypass the intro screen altogether I would think.

**Comment by timbot1789** [comment]
@xscottxbrownx correct, we are able to do that right now. If you leave the site without logging out and then later return (within a certain amount of time), you will automatically log back in. We've actually had that feature for a while, but this fixes some things and makes the refresh a bit faster.

The more significant change is the removal of the hash router.

I think I've updated the PR description since your comment to make things a bit clearer.

**Comment by leekahung** [comment]
Hey @timbot1789, just want to check in on this PR. After this updating this, are there plans to migrate hosting services from GitHub Pages?

If we are, I think I know of a good alternative. I've been using Netlify before for some of my personal projects and it seems to work great with BrowserRouter and integrates well with GitHub repositories.

We just need to take into account Netlify specific files like `netlify.toml` for setting things up, but it should be relatively simple from my experience with that platform. Let me know what you think. Think we could speak with Hugh about getting Netlify integrated to our project if we decide to pursue this route.

File: src/App.jsx [hunk 6]
```diff
@@ -24,18 +23,18 @@ import AppRoutes from './AppRoutes';
...
-    <SessionProvider>
-      <CssBaseline />
-      <ThemeProvider theme={theme}>
-        <Layout ariaLabel="Home Page">
-          <UserDataContextProvider>
+  <SessionProvider restorePreviousSession>
+    <CssBaseline />
+    <ThemeProvider theme={theme}>
+      <UserDataContextProvider>
+        <BrowserRouter>
```
Comment (leekahung) [inline src/App.jsx:29]: Just a bit curious, is there a reason for moving the Router from the uppermost later to just before Layout compared to before where the Router is wrapping around everything?

Comment (timbot1789) [reply src/App.jsx:29]: No strong reason, just thought it looked cleaner. Based the idea off of bulletproof react:

https://github.com/alan2207/bulletproof-react/tree/master

Which just contains a lot of good react programming patterns, and does the routes and provider structure that way.

**Comment by timbot1789** [comment]
> Hey @timbot1789, just want to check in on this PR. After this updating this, are there plans to migrate hosting services from GitHub Pages?
> 
> If we are, I think I know of a good alternative. I've been using Netlify before for some of my personal projects and it seems to work great with BrowserRouter and integrates well with GitHub repositories.
> 
> We just need to take into account Netlify specific files like `netlify.toml` for setting things up, but it should be relatively simple from my experience with that platform. Let me know what you think. Think we could speak with Hugh about getting Netlify integrated to our project if we decide to pursue this route.

Yeah I've been looking at a few different hosting services. Whatever hosting service we go with (or if we need one at all) is something we can talk about in a dev meeting. Netlify is definitely a good option.

Haven't been pushing this because there's a bug in the PR now that I can't track down. The routing works when I run the dev server, but fails when I do the production build and preview it with `npm run preview`. It seems to be an issue with the inrupt auth library, but not 100% sure.

**Comment by timbot1789** [comment]
@leekahung I did some more research, and to clarify, the delay in the PR is due to this bug in the `solid-client-authn-browser`: https://github.com/inrupt/solid-client-authn-js/issues/3001

Basically, the most recent versions of the authn library aren't compatible with vite, and break when passed through `vite build`. I'll investigate some workarounds, but really it would need to be fixed in the library itself.

File: public/_redirects [hunk 7]
```diff
@@ -0,0 +1 @@
+/* /index.html 200
```
Comment (timbot1789) [inline public/_redirects]: This is a routing config for netlify. Without this, reroutes don't work properly when the code is deployed there.

File: src/index.jsx [hunk 8]
```diff
@@ -8,8 +8,4 @@ import App from './App';
 import './style.css';
 
 const root = ReactDOM.createRoot(document.getElementById('root'));
-root.render(
-  <React.StrictMode>
```
Comment (leekahung) [inline src/index.jsx]: I think we can keep `React.StrictMode` to the build. Think, it'll be useful to have it check for any unexpected side-effects during development. Shouldn't affect production.

Comment (timbot1789) [reply src/index.jsx]: Woops forgot I removed that. I'll add it back in.

Comment (timbot1789) [reply src/index.jsx]: Fixed

File: netlify.toml [hunk 9]
```diff
@@ -0,0 +1,4 @@
+[[redirects]]
```
Comment (timbot1789) [inline netlify.toml]: This is a netlify config file. The `[[redirects]]` settings tell the netlify server to let react handle the routing.

File: test/components/NavBar/__snapshots__/OidcLoginComponent.test.jsx.snap [hunk 10]
```diff
@@ -1,108 +0,0 @@
-// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html
```
Comment (timbot1789) [inline test/components/NavBar/__snapshots__/OidcLoginComponent.test.jsx.snap:1]: As we are using material UI more extensively now, these snapshots don't seem to be as useful in testing as I expected them to be. The way material UI writes CSS (composing a series of simple classes) means very small changes to components results in extreme, indecipherable changes to the snapshot. Both Material UI and react testing library also advise against snapshot testing, so I'm removing this snapshot and just using react testing library to look at the elements we're interested in directly.

File: netlify.toml
Comment (leekahung) [inline netlify.toml:1]: I'm not sure if it's needed, but should we have a [build] configuration for the build command with Vite like with `npm run vite:build`?

Also for environment variables, I think there's a way to establish what we have here over to Netlify.

File: vite.config.js [hunk 11]
```diff
@@ -1,17 +1,40 @@
...
   plugins: [
     react(),
     visualizer({
-      filename: 'bundleStats.html',
-      open: true
+      filename: 'bundleStats.html'
     })
   ],
+  resolve:{
+    alias:[
```
Comment (leekahung) [inline vite.config.js:26]: I do like we're going to swap over to using alias for some of our imports. Helps cuts down the lines of text we might get from typing the relative pathing to certain files/directories.

File: netlify.toml
Comment (timbot1789) [reply netlify.toml:1]: We can, but you're right we don't need to. If we don't give it other config, Netlify automatically runs `npm run build` for us and hosts the output. That's all our build process is at the moment. If we ever need a more complex build process than that (which I don't expect we ever will), we can tell netlify to do that here. 

And yes we can add env variables in the netlify terminal, like we do now for code pages. That's the best place to add them

File: vite.config.js (hunk 11 above)
Comment (timbot1789) [reply vite.config.js:26]: Yeah, and if we ever need more, we can just add them to the lists here and in eslint.rc. It made editing all the places we used `useSession` a hell of a lot easier cuz I could just do a regex to replace `@inrupt/solid-react-ui` with `@hooks` :)

Comment (leekahung) [reply vite.config.js:26]: Haha, yeah 😆 

I'm good with this.

**Review by leekahung — COMMENTED** [review]
I believe I've gone over all the files. The underlying code seems to be working fine for redirects and refreshing.

One thing I've noticed is that the auto-populate for adding webId to clients is gone in the production build (things work as intended on the local build on my machine).

https://github.com/codeforpdx/PASS/assets/14917816/ab03c147-0271-496b-9901-c2692d8f01fc

Should be fixed as soon as we got the environment variables set up in Netlify. Outside of that, I think I'm ready to approve this. Think it's ready for our switch from GitHub Pages to another hosting service.

**Comment by timbot1789** [comment]
@leekahung Sorry, I should clarify about the netlify. That's just a demo environment that I set up to show that the code will work in a certain environment. It's not a final PASS prod environment. I mean to compare it to github pages. If you push this branch to github pages and test it there, it will fail.

That netlify environment isn't PASS'S. It's my own on my own account. It's not even using code taken from the PASS repo. I don't have permission to have netlify deploy off of PASS itself; github blocks my connection request. Instead, I made an identical fork and pushed it up.

If we want to use netlify as a prod environment, we'll need to set it up and configure it outside this PR. I just want to show that we _can_ use it.

**Comment by leekahung** [comment]
Ah okay. No worries. I believe we could get @htharker42 to help us with getting the service set up at least with regards to GitHub integration?

**Comment by timbot1789** [comment]
> Ah okay. No worries. I believe we could get @htharker42 to help us with getting the service set up at least with regards to GitHub integration?

Yeah we can talk about it in a dev meeting

File: package.json [hunk 12]
```diff
@@ -54,6 +53,7 @@
     "last 3 edge versions"
   ],
   "devDependencies": {
+    "@rollup/plugin-node-resolve": "^15.1.0",
```
Comment (timbot1789) [inline package.json:56]: These 3 dev dependencies are used to get the `@hooks`, `@components`, etc... mappings working. This one runs on the prod build, the eslint ones run in eslint.

**Review by leekahung — APPROVED** [review]
This is fine. I'm okay with merging this.

**Comment by leekahung** [comment]
Ah right, we'll have to adjust to using Heroku after speaking with Hugh. I'm open to reviewing this again after Heroku related configs are set up.

**Comment by timbot1789** [comment]
> Ah right, we'll have to adjust to using Heroku after speaking with Hugh. I'm open to reviewing this again after Heroku related configs are set up.

I'd like to get this merged as soon as possible. I think there are useful changes here that other developers would benefit from having access to. And the current codebase is already beginning to diverge from the code in this PR. We can put up the heroku config in a separate PR once Hugh gets done setting that up. It's not super complex. I wrote one when I was looking at Heroku for this PR and it's only a few lines at most. If Hugh doesn't get back to us before the next time we want to deploy to prod, I can just put a fork of the project up on my own heroku account and we can use that for a little while.

**Comment by leekahung** [comment]
> > Ah right, we'll have to adjust to using Heroku after speaking with Hugh. I'm open to reviewing this again after Heroku related configs are set up.
> 
> I'd like to get this merged as soon as possible. I think there are useful changes here that other developers would benefit from having access to. And the current codebase is already beginning to diverge from the code in this PR. We can put up the heroku config in a separate PR once Hugh gets done setting that up. It's not super complex. I wrote one when I was looking at Heroku for this PR and it's only a few lines at most. If Hugh doesn't get back to us before the next time we want to deploy to prod, I can just put a fork of the project up on my own heroku account and we can use that for a little while.

I'm good to merge as is now if you think it's ready. If there's other devs you think should need to review the code, I'll wait for their approval first.

**Review by xscottxbrownx — APPROVED** [review]
Went over this PR in discord VC - I had already looked over most files, but was a little hung up on the SessionContext and useSession hook, but good now.

**Review by leekahung — APPROVED** [review]

**Pull request merged**